File: modules/01b-ml-foundations/py/01_linear_regression.py
"""
Task 1 🟡 — Linear regression: the normal equation AND gradient descent.

What you'll learn:
  - The closed-form "normal equation" solution: w = (XᵀX)^{-1} Xᵀy
  - How to solve it stably with np.linalg.solve (never form the inverse)
  - Gradient descent for MSE: the same answer, found iteratively
  - R² (coefficient of determination) as a scale-free goodness-of-fit score

The math (README derives each step):

  Model:            ŷ = X w         (bias absorbed as a column of 1s in X)
  Loss (MSE):       L(w) = (1/N) · ||X w - y||²
  Gradient of MSE:  ∇L  = (2/N) · Xᵀ (X w - y)
  Setting ∇L = 0 :  XᵀX w = Xᵀ y   →   the "normal equation"

  Gradient-descent update:   w ← w - lr · ∇L

  R²:   1 - SS_res / SS_tot ,  where
        SS_res = Σ (y_i - ŷ_i)²   (what the model still gets wrong)
        SS_tot = Σ (y_i - ȳ)²     (variance around the mean baseline)
        R² = 1 → perfect ; R² = 0 → no better than predicting the mean.

You implement the four core numpy functions (normal_equation, predict,
mse_loss, gradient_step). Everything else — synthetic data, standardisation,
the training loop, and the R² report — is provided and runnable.

How to run:
  uv run python modules/01b-ml-foundations/py/01_linear_regression.py
"""

from __future__ import annotations

import numpy as np

SEED = 7
N = 200  # samples
D = 3  # real features (a bias column is prepended → design matrix is N×(D+1))


# ---------------------------------------------------------------------------
# Synthetic data  (provided — do not edit)
# ---------------------------------------------------------------------------


def make_data() -> tuple[np.ndarray, np.ndarray, np.ndarray, float]:
    """
    Generate  y = X_raw @ w_true + b_true + gaussian_noise.

    Returns:
      X_raw  : (N, D) raw feature matrix (unstandardised)
      y      : (N,)   targets
      w_true : (D,)   the weights we secretly used
      b_true : float  the intercept we secretly used
    """
    rng = np.random.default_rng(SEED)
    X_raw = rng.normal(0.0, 1.0, size=(N, D))
    w_true = np.array([2.0, -3.0, 0.5])
    b_true = 4.0
    noise = rng.normal(0.0, 0.5, size=N)
    y = X_raw @ w_true + b_true + noise
    return X_raw, y, w_true, b_true


def standardize(X: np.ndarray) -> np.ndarray:
    """Zero-mean, unit-variance per column. Keeps gradient descent well-scaled."""
    mean = X.mean(axis=0, keepdims=True)
    std = X.std(axis=0, keepdims=True)
    std[std == 0] = 1.0
    return (X - mean) / std


def add_bias_column(X: np.ndarray) -> np.ndarray:
    """Prepend a column of 1s so w[0] plays the role of the intercept b."""
    ones = np.ones((X.shape[0], 1), dtype=X.dtype)
    return np.hstack([ones, X])


# ---------------------------------------------------------------------------
# Core functions — YOU implement these four
# ---------------------------------------------------------------------------


def normal_equation(X: np.ndarray, y: np.ndarray) -> np.ndarray:
    """
    Closed-form least squares:  w = (XᵀX)^{-1} Xᵀ y.

    IMPORTANT: never compute the inverse explicitly (np.linalg.inv is slower and
    numerically worse). Instead assemble the linear system  (XᵀX) w = Xᵀy  and
    hand it to np.linalg.solve, which does an LU factorisation.

    X here already includes the bias column (shape N×(D+1)); the returned w has
    length D+1, with w[0] the intercept.

    TODO: implement.
      1. A = X.T @ X            # shape (D+1, D+1)
      2. b = X.T @ y            # shape (D+1,)
      3. return np.linalg.solve(A, b)
    """
    # TODO: implement the normal equation via np.linalg.solve
    raise NotImplementedError("TODO: implement normal_equation()")


def predict(X: np.ndarray, w: np.ndarray) -> np.ndarray:
    """
    Linear prediction:  ŷ = X @ w.

    X includes the bias column, so no separate intercept term is needed.

    TODO: implement (one line: X @ w).
    """
    # TODO: implement the linear prediction
    raise NotImplementedError("TODO: implement predict()")


def mse_loss(X: np.ndarray, y: np.ndarray, w: np.ndarray) -> float:
    """
    Mean squared error:  L = (1/N) · Σ (ŷ_i - y_i)².

    TODO: implement.
      1. resid = predict(X, w) - y
      2. return float(np.mean(resid ** 2))
    """
    # TODO: implement mean squared error
    raise NotImplementedError("TODO: implement mse_loss()")


def gradient_step(X: np.ndarray, y: np.ndarray, w: np.ndarray, lr: float) -> np.ndarray:
    """
    One gradient-descent step for MSE.

    Gradient:   ∇L = (2/N) · Xᵀ (X w - y)
    Update:     w_new = w - lr · ∇L

    Return the UPDATED weight vector (do not mutate the input in place — the
    training loop reassigns w = gradient_step(...)).

    TODO: implement.
      1. N = X.shape[0]
      2. resid = predict(X, w) - y          # shape (N,)
      3. grad  = (2.0 / N) * (X.T @ resid)  # shape (D+1,)
      4. return w - lr * grad
    """
    # TODO: implement one gradient-descent update
    raise NotImplementedError("TODO: implement gradient_step()")


# ---------------------------------------------------------------------------
# R²  (provided — uses your predict())
# ---------------------------------------------------------------------------


def r_squared(X: np.ndarray, y: np.ndarray, w: np.ndarray) -> float:
    """R² = 1 - SS_res / SS_tot. 1.0 is perfect, 0.0 is mean-baseline."""
    y_hat = predict(X, w)
    ss_res = float(np.sum((y - y_hat) ** 2))
    ss_tot = float(np.sum((y - y.mean()) ** 2))
    return 1.0 - ss_res / ss_tot


# ---------------------------------------------------------------------------
# Harness  (provided — do not edit)
# ---------------------------------------------------------------------------


def main() -> None:
    print("Task 1 — Linear regression: normal equation vs gradient descent\n")

    X_raw, y, w_true, b_true = make_data()
    print(f"  Data: N={N}, D={D}")
    print(f"  True weights: {w_true}   True bias: {b_true}")

    # Standardise features, then add the bias column.
    X = add_bias_column(standardize(X_raw))
    print(f"  Design matrix (with bias col): {X.shape}\n")

    # ── Closed form ──────────────────────────────────────────────────────────
    print("[1/2] Normal equation (closed form)...")
    w_normal = normal_equation(X, y)
    print(f"  w_normal = {np.round(w_normal, 4)}")
    print(f"  MSE      = {mse_loss(X, y, w_normal):.4f}")
    print(f"  R²       = {r_squared(X, y, w_normal):.4f}\n")

    # ── Gradient descent ─────────────────────────────────────────────────────
    print("[2/2] Gradient descent...")
    rng = np.random.default_rng(SEED)
    w_gd = rng.normal(0.0, 0.1, size=X.shape[1])  # small random init
    lr = 0.1
    epochs = 300

    loss_history: list[float] = []
    for _ in range(epochs):
        loss_history.append(mse_loss(X, y, w_gd))
        w_gd = gradient_step(X, y, w_gd, lr)

    final_loss = mse_loss(X, y, w_gd)

    # Is the loss monotonically decreasing over the first 30 epochs?
    first30 = loss_history[:30]
    monotonic = all(first30[i + 1] <= first30[i] + 1e-9 for i in range(len(first30) - 1))

    # How close did GD get to the closed-form optimum?
    dist = float(np.linalg.norm(w_gd - w_normal))
    r2_gd = r_squared(X, y, w_gd)

    for e in (1, 10, 50, 150, epochs):
        print(f"  epoch {e:>4}: MSE={loss_history[e - 1]:.4f}")
    print(f"\n  w_gd     = {np.round(w_gd, 4)}")
    print(f"  final MSE (GD)          = {final_loss:.4f}")
    print(f"  ||w_gd - w_normal||     = {dist:.4f}")
    print(f"  R² (GD)                 = {r2_gd:.4f}")
    print(f"  loss monotone (first 30)= {monotonic}")

    # ── Acceptance checks ────────────────────────────────────────────────────
    print("\nAcceptance:")
    ok_converge = dist < 0.1
    ok_monotone = monotonic
    ok_r2 = r2_gd > 0.9
    print(
        f"  [{'x' if ok_converge else ' '}] GD converges to normal eq  (||Δw|| = {dist:.4f} < 0.1)"
    )
    print(f"  [{'x' if ok_monotone else ' '}] MSE decreases monotonically over first 30 epochs")
    print(f"  [{'x' if ok_r2 else ' '}] R² > 0.9  (R² = {r2_gd:.4f})")

    if ok_converge and ok_monotone and ok_r2:
        print("\n  All acceptance checks passed.")
    else:
        print("\n  Some checks failed — revisit your implementation.")


if __name__ == "__main__":
    main()

File: modules/01b-ml-foundations/ts/01-linear-regression.ts
/**
 * Task 1 🟡 — Linear regression: the normal equation AND gradient descent.
 *
 * What you'll learn:
 *   - The closed-form "normal equation" solution: w = (XᵀX)^{-1} Xᵀy
 *   - How to solve it via a linear-system solve (never form the inverse)
 *   - Gradient descent for MSE: the same answer, found iteratively
 *   - R² (coefficient of determination) as a scale-free goodness-of-fit score
 *
 * The math (README derives each step):
 *
 *   Model:            ŷ = X w         (bias absorbed as a column of 1s in X)
 *   Loss (MSE):       L(w) = (1/N) · ||X w - y||²
 *   Gradient of MSE:  ∇L  = (2/N) · Xᵀ (X w - y)
 *   Setting ∇L = 0 :  XᵀX w = Xᵀ y   →   the "normal equation"
 *
 *   Gradient-descent update:   w ← w - lr · ∇L
 *
 *   R²:   1 - SS_res / SS_tot
 *
 * A Gaussian-elimination `solve(A, b)` is PROVIDED in the harness. You implement
 * the four core functions using plain arrays only (no math libraries):
 * normalEquation (assemble XᵀX and Xᵀy, then call solve), predict, mseLoss,
 * gradientStep. Everything else is provided and runnable.
 *
 * How to run:
 *   pnpm tsx modules/01b-ml-foundations/ts/01-linear-regression.ts
 */

const SEED = 7;
const N = 200; // samples
const D = 3; // real features (a bias column is prepended → design matrix is N×(D+1))

// ---------------------------------------------------------------------------
// Seeded RNG (provided) — LCG + Box-Muller for reproducible Gaussians
// ---------------------------------------------------------------------------

function makeRng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    // Numerical Recipes LCG
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296; // uniform in [0, 1)
  };
}

function makeGaussian(seed: number): () => number {
  const u = makeRng(seed);
  return () => {
    // Box-Muller: two uniforms → one standard normal
    const u1 = u() + 1e-12;
    const u2 = u();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  };
}

// ---------------------------------------------------------------------------
// Matrix / vector helpers (provided)
// ---------------------------------------------------------------------------

/** A (N×K) @ B (K×M) → (N×M). */
function matMul(A: number[][], B: number[][]): number[][] {
  const n = A.length,
    k = A[0].length,
    m = B[0].length;
  const C = Array.from({ length: n }, () => new Array(m).fill(0) as number[]);
  for (let i = 0; i < n; i++)
    for (let p = 0; p < k; p++)
      for (let j = 0; j < m; j++) C[i][j] += A[i][p] * B[p][j];
  return C;
}

/** A (N×M) @ v (M) → (N). */
function matVec(A: number[][], v: number[]): number[] {
  return A.map((row) => row.reduce((acc, a, j) => acc + a * v[j], 0));
}

/** Transpose A (N×M) → (M×N). */
function transpose(A: number[][]): number[][] {
  const n = A.length,
    m = A[0].length;
  const T = Array.from({ length: m }, () => new Array(n).fill(0) as number[]);
  for (let i = 0; i < n; i++) for (let j = 0; j < m; j++) T[j][i] = A[i][j];
  return T;
}

/**
 * Solve the linear system A x = b for x via Gaussian elimination with partial
 * pivoting. A is (K×K), b is (K). Returns x of length K. (Provided so you don't
 * have to build a solver from scratch — your job is to ASSEMBLE the system.)
 */
function solve(A: number[][], b: number[]): number[] {
  const n = A.length;
  // Build an augmented copy [A | b].
  const M = A.map((row, i) => [...row, b[i]]);
  for (let col = 0; col < n; col++) {
    // Partial pivot: find the row with the largest magnitude in this column.
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivot][col])) pivot = r;
    }
    [M[col], M[pivot]] = [M[pivot], M[col]];
    const piv = M[col][col];
    if (Math.abs(piv) < 1e-12) throw new Error("Matrix is singular");
    // Eliminate this column from every other row.
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = M[r][col] / piv;
      for (let c = col; c <= n; c++) M[r][c] -= factor * M[col][c];
    }
  }
  // Back-substitute (matrix is now diagonal).
  return M.map((row, i) => row[n] / row[i]);
}

// ---------------------------------------------------------------------------
// Synthetic data (provided)
// ---------------------------------------------------------------------------

function makeData(): { Xraw: number[][]; y: number[]; wTrue: number[]; bTrue: number } {
  const g = makeGaussian(SEED);
  const Xraw = Array.from({ length: N }, () => Array.from({ length: D }, () => g()));
  const wTrue = [2.0, -3.0, 0.5];
  const bTrue = 4.0;
  const noise = makeGaussian(SEED + 99);
  const y = Xraw.map(
    (row) => row.reduce((acc, x, j) => acc + x * wTrue[j], bTrue) + 0.5 * noise(),
  );
  return { Xraw, y, wTrue, bTrue };
}

/** Zero-mean, unit-variance per column. */
function standardize(X: number[][]): number[][] {
  const cols = X[0].length;
  const mean = new Array(cols).fill(0);
  const std = new Array(cols).fill(0);
  for (const row of X) for (let j = 0; j < cols; j++) mean[j] += row[j] / X.length;
  for (const row of X)
    for (let j = 0; j < cols; j++) std[j] += (row[j] - mean[j]) ** 2 / X.length;
  for (let j = 0; j < cols; j++) std[j] = Math.sqrt(std[j]) || 1;
  return X.map((row) => row.map((v, j) => (v - mean[j]) / std[j]));
}

/** Prepend a column of 1s so w[0] is the intercept. */
function addBiasColumn(X: number[][]): number[][] {
  return X.map((row) => [1, ...row]);
}

// ---------------------------------------------------------------------------
// Core functions — YOU implement these four (plain arrays only)
// ---------------------------------------------------------------------------

/**
 * Closed-form least squares: w = (XᵀX)^{-1} Xᵀ y.
 *
 * Never invert a matrix. Instead assemble the system (XᵀX) w = Xᵀy and call
 * the provided solve(A, b).
 *
 * X already includes the bias column (N×(D+1)); the returned w has length D+1,
 * with w[0] the intercept.
 *
 * TODO: implement.
 *   1. const Xt = transpose(X);         // (D+1)×N
 *   2. const A  = matMul(Xt, X);        // (D+1)×(D+1)  == XᵀX
 *   3. const bb = matVec(Xt, y);        // (D+1)        == Xᵀy
 *   4. return solve(A, bb);
 */
function normalEquation(X: number[][], y: number[]): number[] {
  // TODO: assemble XᵀX and Xᵀy, then call solve()
  throw new Error("TODO: implement normalEquation()");
}

/**
 * Linear prediction: ŷ = X @ w. X includes the bias column.
 *
 * TODO: implement (return matVec(X, w)).
 */
function predict(X: number[][], w: number[]): number[] {
  // TODO: implement the linear prediction
  throw new Error("TODO: implement predict()");
}

/**
 * Mean squared error: L = (1/N) · Σ (ŷ_i - y_i)².
 *
 * TODO: implement.
 *   1. const yhat = predict(X, w);
 *   2. sum the squared residuals (yhat[i] - y[i])²
 *   3. return the sum divided by N
 */
function mseLoss(X: number[][], y: number[], w: number[]): number {
  // TODO: implement mean squared error
  throw new Error("TODO: implement mseLoss()");
}

/**
 * One gradient-descent step for MSE.
 *
 * Gradient:   ∇L = (2/N) · Xᵀ (X w - y)
 * Update:     w_new = w - lr · ∇L
 *
 * Return a NEW weight array (do not mutate the input; the loop reassigns w).
 *
 * TODO: implement.
 *   1. const n = X.length;
 *   2. const resid = predict(X, w).map((p, i) => p - y[i]);   // length N
 *   3. const grad  = matVec(transpose(X), resid).map(g => (2 / n) * g); // length D+1
 *   4. return w.map((wj, j) => wj - lr * grad[j]);
 */
function gradientStep(X: number[][], y: number[], w: number[], lr: number): number[] {
  // TODO: implement one gradient-descent update
  throw new Error("TODO: implement gradientStep()");
}

// ---------------------------------------------------------------------------
// R² (provided — uses your predict())
// ---------------------------------------------------------------------------

function rSquared(X: number[][], y: number[], w: number[]): number {
  const yhat = predict(X, w);
  const yMean = y.reduce((a, b) => a + b, 0) / y.length;
  let ssRes = 0;
  let ssTot = 0;
  for (let i = 0; i < y.length; i++) {
    ssRes += (y[i] - yhat[i]) ** 2;
    ssTot += (y[i] - yMean) ** 2;
  }
  return 1 - ssRes / ssTot;
}

// ---------------------------------------------------------------------------
// Harness (provided)
// ---------------------------------------------------------------------------

function round(v: number[], places = 4): number[] {
  const f = 10 ** places;
  return v.map((x) => Math.round(x * f) / f);
}

function main(): void {
  console.log("Task 1 — Linear regression: normal equation vs gradient descent\n");

  const { Xraw, y, wTrue, bTrue } = makeData();
  console.log(`  Data: N=${N}, D=${D}`);
  console.log(`  True weights: [${wTrue}]   True bias: ${bTrue}`);

  const X = addBiasColumn(standardize(Xraw));
  console.log(`  Design matrix (with bias col): ${X.length}×${X[0].length}\n`);

  // ── Closed form ────────────────────────────────────────────────────────────
  console.log("[1/2] Normal equation (closed form)...");
  const wNormal = normalEquation(X, y);
  console.log(`  wNormal = [${round(wNormal)}]`);
  console.log(`  MSE     = ${mseLoss(X, y, wNormal).toFixed(4)}`);
  console.log(`  R²      = ${rSquared(X, y, wNormal).toFixed(4)}\n`);

  // ── Gradient descent ───────────────────────────────────────────────────────
  console.log("[2/2] Gradient descent...");
  const gInit = makeGaussian(SEED);
  let wGd = Array.from({ length: X[0].length }, () => 0.1 * gInit());
  const lr = 0.1;
  const epochs = 300;

  const lossHistory: number[] = [];
  for (let epoch = 0; epoch < epochs; epoch++) {
    lossHistory.push(mseLoss(X, y, wGd));
    wGd = gradientStep(X, y, wGd, lr);
  }

  const finalLoss = mseLoss(X, y, wGd);

  // Monotone over first 30 epochs?
  const first30 = lossHistory.slice(0, 30);
  let monotonic = true;
  for (let i = 0; i < first30.length - 1; i++) {
    if (first30[i + 1] > first30[i] + 1e-9) monotonic = false;
  }

  // Distance to the closed-form optimum.
  let dist = 0;
  for (let j = 0; j < wGd.length; j++) dist += (wGd[j] - wNormal[j]) ** 2;
  dist = Math.sqrt(dist);
  const r2Gd = rSquared(X, y, wGd);

  for (const e of [1, 10, 50, 150, epochs]) {
    console.log(
      `  epoch ${String(e).padStart(4)}: MSE=${lossHistory[e - 1].toFixed(4)}`,
    );
  }
  console.log(`\n  wGd      = [${round(wGd)}]`);
  console.log(`  final MSE (GD)          = ${finalLoss.toFixed(4)}`);
  console.log(`  ||wGd - wNormal||       = ${dist.toFixed(4)}`);
  console.log(`  R² (GD)                 = ${r2Gd.toFixed(4)}`);
  console.log(`  loss monotone (first 30)= ${monotonic}`);

  // ── Acceptance checks ──────────────────────────────────────────────────────
  console.log("\nAcceptance:");
  const okConverge = dist < 0.1;
  const okMonotone = monotonic;
  const okR2 = r2Gd > 0.9;
  console.log(
    `  [${okConverge ? "x" : " "}] GD converges to normal eq  (||Δw|| = ${dist.toFixed(4)} < 0.1)`,
  );
  console.log(
    `  [${okMonotone ? "x" : " "}] MSE decreases monotonically over first 30 epochs`,
  );
  console.log(`  [${okR2 ? "x" : " "}] R² > 0.9  (R² = ${r2Gd.toFixed(4)})`);

  if (okConverge && okMonotone && okR2) {
    console.log("\n  All acceptance checks passed.");
  } else {
    console.log("\n  Some checks failed — revisit your implementation.");
  }
}

main();

// This file has no imports; the empty export marks it as an ES module so its
// top-level names stay file-scoped (keeps `pnpm typecheck` happy across modules).
export {};
